Initial Commit for player custom quest assignment

whimsicaldev opened the pull request:
Please review the email sending for assign custom quest functionality

whimsicaldev committed:
PlayerProfile: Assign Custom Quest

initial commit for the player assigne custom quest

whimsicaldev committed:
Merge branch 'master' into ge-branched-origin

whimsicaldev committed:
PlayerProfile: Assign Custom Quest

updating changes for the email notification

whimsicaldev committed:
NO-REF

updated the typescript version since angular cli fails when the version is not met

whimsicaldev committed:
PlayerView: Assign Customer Quest

Updating the email body for quest assignment

whimsicaldev committed:
NO-REF

reverted to original label

whimsicaldev committed:
Merge branch 'master' into ge-branched-origin

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -0,0 +1,73 @@
+import { NotifyService } from '../../../core/notify.service';
+import { EmailService } from '../../../core/email.service';
+
+const CategoryList = [
```
we already have QuestCategories in player-quest.d.ts

whimsicaldev committed:
NO-REF

updated files to make lint pass

ben-x-dev commented on src/app/ui/user-profile/user-profile.component.ts:
```diff
@@ -1,17 +1,76 @@
+            this.userServiceSub = this.userService.getUser(playerId).subscribe((player: User) => {
+              this.player = player;
+
+              this.seasonSub = this.seasonService.getEnabledSeason().subscribe((season: Season) => {
```
i think this is not dependent at all on the user info. 
and also to prevent using stale seasonId (where enabled season changes after you've get the id), we need to do this just before the transaction where it is really needed.

whimsicaldev replied:
the team lead is assigning the quest to the user, the current user data loaded is the team lead's data. For the player's quest creation, data needed is the specific user's details.

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -0,0 +1,73 @@
+      seasonId: this.season.id,
+      xp: 10,
+      playerEmail: this.user.email
+    } as PlayerQuest;
```
lets refrain from placing logic inside the constructor. please move these on ngOnInit instead

whimsicaldev committed:
NO-REF

fix for ng build --prod issue

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -0,0 +1,73 @@
+  templateUrl: './add-quest-dialog.component.html',
+  styleUrls: ['./add-quest-dialog.component.scss']
+})
+export class AddQuestDialogComponent implements OnInit {
```
due to limitations in MS api, we cannot whitelist/register dynamic urls to access their API. Therefore it might be best to turn this into a page component and add a static route for it. I think it will also benefit us for future improvements in this area where we may need occupy more space that may not fit in the modal.

whimsicaldev committed:
NO-REF

initial update on the implementation based on the review points

whimsicaldev committed:
NO-REF

shadowed variable for lint

whimsicaldev committed:
NO-REF

final update for send email updates

whimsicaldev committed:
NO-REF

request for team lead email is updated

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -1,73 +1,85 @@
+      this.playerQuest.seasonId = season.id;
+
+      this.playerQuestService.assignPlayerQuest(this.playerQuest).then((docRef: DocumentReference) => {
+        docRef.get().then((data) => {
```
I'm pretty sure you don't need to do `docRef.get()` here since going inside `then()` means request is successful, otherwise it will throw an error which can be caught via `catch()`.
We need to be prudent in making requests so we don't go beyond firebase's free tier limit

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -1,73 +1,85 @@
-      xp: 10,
-      playerEmail: this.user.email
-    } as PlayerQuest;
+    this.lead = data.lead;
```
these two can also be move to `ngOnInit` since you can get `data` via `this.data`

ben-x-dev commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,45 @@
+/* 
```
please add this to .gitignore as well

ben-x-dev commented on package.json:
```diff
@@ -15,60 +15,72 @@
-    "firebase": "^5.5.6",
+    "@nguniversal/module-map-ngfactory-loader": "^7.1.0",
+    "@schematics/angular": "^7.3.3",
+    "buffer": "^5.2.1",
```
why do we need this?

ben-x-dev commented on package.json:
```diff
@@ -15,60 +15,72 @@
+    "ws": "^5.2.2",
     "xmlhttprequest": "^1.8.0"
+  },
+  "browser": {
```
what is this for?

ben-x-dev commented on sendgrid.env:
```diff
@@ -0,0 +1 @@
+"export SENDGRID_API_KEY='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'" 
```
i assume this is from before when you were trying out sending emails? so we don't need this anymore right?

ben-x-dev commented on src/app/app-routing.module.ts:
```diff
@@ -14,6 +14,7 @@ import { UploadPageComponent } from './uploads/upload-page/upload-page.component
 const routes: Routes = [
   { path: '', component: HomePageComponent },
   { path: 'login', component: UserLoginComponent },
+  { path: 'profile/:playerId', component: UserProfileComponent, canActivate: [AuthGuard]},
```
please remove

ben-x-dev commented on src/app/app.component.html:
```diff
@@ -31,7 +31,7 @@
         </a>
         <a mat-list-item [routerLinkActive]="['active-menu']"
-          routerLink="/profile">
+          routerLink="/profile/">
```
lets just drop the last slash, ms api is very strict on urls

ben-x-dev commented on src/app/ui/player-quest/player-quest.d.ts:
```diff
@@ -28,5 +29,6 @@ interface PlayerQuest {
   questName: string;
   source: string;  // e.g. LMS
   required: boolean; // true = 10 points granted, else 5 points
+  xp: number;
```
hmm don't think we need xp here, having this may cause future conflicts.
required is still our single source of truth with regards to xp points awarded.

ben-x-dev commented on src/app/ui/user-profile/user-profile.component.html:
```diff
@@ -1,43 +1,73 @@
-      <button mat-stroked-button color="accent" (click)="logout()">Logout</button>
-    </mat-card-actions>
-  </mat-card>
+  <div *ngIf="true then profile else null">
```
this is always true so lets just remove this please

ben-x-dev commented on src/app/ui/user-profile/user-profile.component.html:
```diff
@@ -1,43 +1,73 @@
+  <div class="box" *ngIf="user.isLead && !player">
+    <mat-card class="example-card">
+      <mat-card-actions>
+        <button mat-stroked-button color="primary" (click)="viewUserProfile('Y2FlYmU0YmItMTkyYi00NGQwLTljYmQtYjEyODJmNjVjZTdl.NDU3ZDU2ODUtMDQ2Ny00ZDA1LWIyM2ItOGY4MTdhZGRhNDdj')">View Profile</button> 
```
1 problem with this approach is: team lead won't be able to go back to his list of team members.
using browser's back button or clicking profile from the sidenav wouldn't work.
Perhaps we should push to location history to programmatically set a location change e.g. /profile/member ?

whimsicaldev replied:
updated the view, if you click the profile now, the team lead will be navigated away

ben-x-dev commented on src/app/ui/user-profile/user-profile.component.ts:
```diff
@@ -1,17 +1,57 @@
-export class UserProfileComponent {
+export class UserProfileComponent implements OnInit, OnDestroy {
+  private player: User;
+  private userServiceSub: any;
```
please use Subscription type for subscriptions

whimsicaldev committed:
NO-REF

committing the review points

whimsicaldev committed:
NO-REF

fix for lint issue

whimsicaldev committed:
NO-REF

whimsicaldev committed:
NO-REF

profile link to remove the forward slash

whimsicaldev committed:
NO-REF

updated the fix for the link issue

whimsicaldev committed:
NO-REF

final touches

whimsicaldev committed:
NO-REF

whimsicaldev committed:
Merge branch 'master' into ge-branched-origin

whimsicaldev committed:
NO-REF

gerarrd commented:
Hey,

The API works if you pass the user id as a request parameter. So I updated the code to use it.

Cheers!
Gê

ben-x-dev commented on .gitignore:
```diff
@@ -43,5 +43,7 @@ testem.log
 .DS_Store
 Thumbs.db
+.vscode/tasks.json
+package-lock.json
```
please don't gitignore this.

ben-x-dev commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,45 @@
+/* 
```
adding this to gitignore means this should no longer be part of the commit.
please remove this file from the commit

ben-x-dev replied:
why is this still here?

ben-x-dev commented on src/app/core/user.service.ts:
```diff
@@ -0,0 +1,25 @@
+  AngularFirestoreCollection,
+  AngularFirestoreDocument
+} from '@angular/fire/firestore';
+import { firestore } from 'firebase/app';
```
please remove all unused imports

ben-x-dev commented on src/app/ui/dialog/add-quest-dialog/add-quest-dialog.component.ts:
```diff
@@ -0,0 +1,47 @@
+  ];
+  public playerQuest: PlayerQuest;
+  public user: User;
+  private lead: User;
```
this property is not used.

ben-x-dev commented on src/app/ui/user-profile/user-profile.component.ts:
```diff
@@ -1,17 +1,93 @@
+    // get the logged-in user data
+    this.userSub = this.auth.user.subscribe((user: User) => {
+      this.user = user;
+      if (this.user.team.lead === this.user.uid) {
```
no need for this here, i'll fix the security rules for getUser to allow only owner and team lead for read access.
It will throw an error if it fails, so you need to catch the observable exception and just show a simple placeholder that the user is not allowed access to this profile.

ben-x-dev reviewed (changes requested):
please remove all `package.json` and `package-lock.json` file changes from this commit. 
lets only commit them if there are new dependencies or scripts that has to be been added.

whimsicaldev committed:
NO-REF

whimsicaldev committed:
NO-REF

reverting the package and package-lock files

whimsicaldev committed:
NO-REF

reverting the changes in the package and package-lock files in the functions directory

whimsicaldev committed:
deleting the file

whimsicaldev committed:
NO-REF

reverting unchaged files

whimsicaldev committed:
Merge branch 'ge-branched-origin' of https://github.com/nbcp/firestarter-leaderboard into ge-branched-origin

whimsicaldev committed:
NO-REF
